worlds/sc2wol/MissionTables.py

```diff
@@ -171,40 +287,103 @@ class FillMission(NamedTuple):
+        SC2Mission.CUTTHROAT.mission_name: MissionInfo(SC2Mission.CUTTHROAT, [MissionConnection(17, SC2Campaign.WOL)], SC2Mission.THE_GREAT_TRAIN_ROBBERY.area),
+        SC2Mission.ENGINE_OF_DESTRUCTION.mission_name: MissionInfo(SC2Mission.ENGINE_OF_DESTRUCTION, [MissionConnection(18, SC2Campaign.WOL)], SC2Mission.ENGINE_OF_DESTRUCTION.area),
+        SC2Mission.MEDIA_BLITZ.mission_name: MissionInfo(SC2Mission.MEDIA_BLITZ, [MissionConnection(19, SC2Campaign.WOL)], SC2Mission.MEDIA_BLITZ.area),
+        SC2Mission.PIERCING_OF_THE_SHROUD.mission_name: MissionInfo(SC2Mission.PIERCING_OF_THE_SHROUD, [MissionConnection(20, SC2Campaign.WOL)], SC2Mission.PIERCING_OF_THE_SHROUD.area),
+        SC2Mission.GATES_OF_HELL.mission_name: MissionInfo(SC2Mission.GATES_OF_HELL, [MissionConnection(12, SC2Campaign.WOL)], SC2Mission.GATES_OF_HELL.area, completion_critical=True),
+        SC2Mission.BELLY_OF_THE_BEAST.mission_name: MissionInfo(SC2Mission.BELLY_OF_THE_BEAST, [MissionConnection(22, SC2Campaign.WOL)], SC2Mission.BELLY_OF_THE_BEAST.area, completion_critical=True),
+        SC2Mission.SHATTER_THE_SKY.mission_name: MissionInfo(SC2Mission.SHATTER_THE_SKY, [MissionConnection(22, SC2Campaign.WOL)], SC2Mission.SHATTER_THE_SKY.area, completion_critical=True),
+        SC2Mission.ALL_IN.mission_name: MissionInfo(SC2Mission.ALL_IN, [MissionConnection(23, SC2Campaign.WOL), MissionConnection(24, SC2Campaign.WOL)], SC2Mission.ALL_IN.area, or_requirements=True, completion_critical=True)
+    },
+    SC2Campaign.PROPHECY: {
+        SC2Mission.WHISPERS_OF_DOOM.mission_name: MissionInfo(SC2Mission.WHISPERS_OF_DOOM, [MissionConnection(9, SC2Campaign.WOL)], SC2Mission.WHISPERS_OF_DOOM.area),
+        SC2Mission.A_SINISTER_TURN.mission_name: MissionInfo(SC2Mission.A_SINISTER_TURN, [MissionConnection(1, SC2Campaign.PROPHECY)], SC2Mission.A_SINISTER_TURN.area),
+        SC2Mission.ECHOES_OF_THE_FUTURE.mission_name: MissionInfo(SC2Mission.ECHOES_OF_THE_FUTURE, [MissionConnection(2, SC2Campaign.PROPHECY)], SC2Mission.A_SINISTER_TURN.area),
+        SC2Mission.IN_UTTER_DARKNESS.mission_name: MissionInfo(SC2Mission.IN_UTTER_DARKNESS, [MissionConnection(3, SC2Campaign.PROPHECY)], SC2Mission.IN_UTTER_DARKNESS.area)
+    }
+}
+
+lookup_id_to_mission: Dict[int, SC2Mission] = {
+    mission.id: mission for mission in SC2Mission
+}
+
+lookup_name_to_mission: Dict[str, SC2Mission] = {
+    mission.mission_name: mission for mission in SC2Mission
+}
+
+lookup_id_to_campaign: Dict[int, SC2Campaign] = {
+    campaign.id: campaign for campaign in SC2Campaign
 }
 
-lookup_id_to_mission: Dict[int, str] = {
-    data.id: mission_name for mission_name, data in vanilla_mission_req_table.items() if data.id}
+
+def get_campaign_missions(campaign: SC2Campaign) -> List[SC2Mission]:
+    return [mission for mission in SC2Mission if mission.campaign == campaign]
```

Comment:
If I'm counting correctly, this is called once for every enabled campaign, once for every disabled campaign, and then once again for every enabled campaign if the mission order is vanilla. So since every generation will iterate the entire mission enum at least once for every campaign, it's much better to use a lookup table for this.

Maybe `vanilla_mission_req_table` can be reused, since it's the same information? Eg.:
```python
campaign_mission_table: Dict[SC2Campaign, Set[SC2Mission]] = {
  campaign: {info.mission for info in mission_reqs.values()}
  for (campaign, mission_reqs) in vanilla_mission_req_table.items()
}
```
But I don't know how future-proof that is.

Reply:
I think that may fail with Emperor's Fall (like All-in but using two separate maps instead). For Vanilla mission order you want a config if Ground Zero or Birds of War variant will take the place. There should be one of them in Vanilla mission order but the other also aligned in the same campaign and when the option says to use the other map, make the switch

Reply:
I think that'd be fine actually - `get_campaign_missions` is only used to (1) exclude disabled campaigns and (2) get included missions to determine max campaign difficulty. So whatever placeholder is in vanilla reqs (I would assume a generic "Emperor's Fall") would suffice for the uses of this code. Since this table is built at compile time, the generation can freely mess with the vanilla req table to switch out missions.

The only thing that wouldn't work is if someone wanted to manually exclude only one version of the map, but I think that would be fine to not support since we have an option for that. If they wanted to exclude both versions they can exclude "Emperor's Fall" instead, which would work with this.

Reply:
Well, this operation isn't called often (just a few times) so it doesn't increase really time complexity much. Therefore I'd prefer an approach that "always works" over "slightly optimized but could cause bugs in the future".

If it'd be called often, yeah, but I don't think it's worth now to risk the potential side-effects.

Reply:
It's called once per campaign minimum and twice per campaign maximum, which if we're already counting SC1/BW campaigns is 12-24 times, for, what, 150-200 missions? I can accept the decision to not base it on the vanilla req table, but then it's still work that doesn't need to be repeated this much.
```python
campaign_mission_table: Dict[SC2Campaign, Set[SC2Mission]] = {
  campaign: set()
  for campaign in SC2Campaign
}
for mission in SC2Mission:
  campaign_mission_table[mission.campaign].add(mission)
```
This less fancy snippet does exactly what the function currently does, except 11+ times less often.

Reply:
Done